### DISCUSSION TITLE
Changes in 2024.03

### DISCUSSION THREAD

**Discussion by ankraft** [ankraft/ACME-oneM2M-CSE#131] [Announcements]
The following changes are in release 2024.03

## New Features
- AddedWebSocket Binding support.  
This binding is fully functional but still experimental because it differs from the oneM2M spec in some points, e.g. the new presence of the *Originator* header when establishing a WebSocket connection. The general functionality is the same, though.  
Contribution to the TS-0020 to adopt this change will be prepared.
  - Added WebSocket binding support for the test suite.
  - Added extra tests cases that test the WebSocket specific behavior for registrations and notifications.
- Added support for *SIM* and *MobileNetwork* ManagementObject specializations.
- Added support for Python 3.12.

## Improvements
- CSE Improvements
  - Allow virtual resources to be referenced in &lt;actions> and &lt;dependencies>.
- Improved Text UI
  - Auto-refresh for diagrams
  - Diagrams now support containers that contain boolean-kind values.
  - Added *Service* menu for resources.
    - Export a resource and its child resources as *curl* commands.
    - Export the *instance* resources of &lt;container> and &lt;timeSeries> container to CSV files.
- Script Interpreter
  - Added "remove-json-attribute", "block", "return-from", and "unwind-protect" functions.

## Fixes
- Disallow &lt;contentInstance> deletion when the parent &lt;container> has the *disallowRetrieval* attribute set to *true*.
- Fixed some incorrect http headers.
- Fixed empty result content in the &lt;REQ> instance when forwarding non-blocking requests to a remote CSE.
- Fixed wrong handling of *Originator* header in forwarded requests, especially when creating new remote resources.
- CSE-relative identifiers in attributes are now correctly converted to SP-relative identifiers when announcing resources.

## Experimental
- The WebSocket could change depending on discussion in the oneM2M working groups.

## Breaking Changes
N/A

**Comment by name-taken23** [comment]
Hi Ankraft,

There seems to be a potential breaking change with the MQTT- Paho have updated to 2.0 and the MQTT client init now requires an extra parameter specifying Apicallback version

https://github.com/eclipse/paho.mqtt.python/releases/tag/v2.0.0

I resolved this by downgrading the mqtt packages on my system
`python3 -m pip install paho-mqtt==1.6.1`

**Reply by ankraft** [reply]
Thanks for reporting. I added issue #147 to handle this.
